python_modules/dagster/dagster/_core/definitions/data_version.py

```diff
@@ -581,23 +580,23 @@ def _get_stale_causes_materialized(self, key: "AssetKeyPartitionKey") -> Iterato
 
     @cached_method
     def _get_stale_root_causes(self, key: "AssetKeyPartitionKey") -> Sequence[StaleCause]:
-        causes = self._get_stale_causes(key=key)
-        root_pairs = sorted([pair for cause in causes for pair in self._gather_leaves(cause)])
-        # After sorting the pairs, we can drop the level and de-dup using an
-        # ordered dict as an ordered set. This will give us unique root causes,
-        # sorted by level.
-        roots: Dict[StaleCause, None] = OrderedDict()
-        for root_cause in [leaf_cause for _, leaf_cause in root_pairs]:
-            roots[root_cause] = None
-        return list(roots.keys())
-
-    # The leaves of the cause tree for an asset are the root causes of its staleness.
-    def _gather_leaves(self, cause: StaleCause, level: int = 0) -> Iterator[Tuple[int, StaleCause]]:
-        if cause.children is None:
-            yield (level, cause)
-        else:
-            for child in cause.children:
-                yield from self._gather_leaves(child, level=level + 1)
+        candidates = self._get_stale_causes(key=key)
+        visited: Set[str] = set()
+        root_causes: List[StaleCause] = []
+        level = 0
+        while candidates:
+            next_candidates: List[StaleCause] = []
+            for cause in candidates:
+                if cause.sort_key not in visited:
+                    if cause.children is None:
+                        root_causes.append(cause)
+                    else:
+                        next_candidates.extend(cause.children)
+                    visited.add(cause.sort_key)  # maybe replace with better key
```

Comment:
Somewhat surprised that using this sort_key as the unique identifier here isn't causing any problems -- I'd expect there to be cases where there are duplicates here (i.e. a parent's code version AND data version have both updated), and in that case it looks like this code would only capture one of those. easy fix would just be to use a tuple that contains the reason/cause category properties in there as well I think, but maybe I'm missing something

Reply:
Ya thats what my comment was getting at. I was curious if I would get any test failures. Can definitely replace the key set up like you are suggesting.